### ISSUE TITLE
Password submission timeout

### ISSUE THREAD

**Issue by stalsma** [thiagobruch/amazon_shopping_list_scrape#3]
I'm trying to get this working using the German Amazon site.  And I've seen it work a few times, but that seems to be the exception.  The username and password get sent, but there's is almost always a timeout while waiting for navigation.

I added a few console.log() to the script to isolate the issue:
```
// const ids = await page.$$eval('[id]', elements => elements.map(el => el.id));
// console.log('IDs found on the page:', ids);
            await page.type('#ap_password', amz_password);
            await page.click('#signInSubmit');
                console.log('filled pwd, and submitting');
            await page.waitForNavigation();
                console.log('submit successful');
```

When it runs, I only see my first log message:
```
filled pwd, and submitting
/app/node_modules/puppeteer-core/lib/cjs/puppeteer/util/Deferred.js:59
            this.#timeoutError = new Errors_js_1.TimeoutError(opts.message);
                                 ^
TimeoutError: Navigation timeout of 60000 ms exceeded
    at new Deferred (/app/node_modules/puppeteer-core/lib/cjs/puppeteer/util/Deferred.js:59:34)
    at Deferred.create (/app/node_modules/puppeteer-core/lib/cjs/puppeteer/util/Deferred.js:21:16)
    at new LifecycleWatcher (/app/node_modules/puppeteer-core/lib/cjs/puppeteer/cdp/LifecycleWatcher.js:69:60)
    at CdpFrame.waitForNavigation (/app/node_modules/puppeteer-core/lib/cjs/puppeteer/cdp/Frame.js:199:29)
    at CdpFrame.<anonymous> (/app/node_modules/puppeteer-core/lib/cjs/puppeteer/util/decorators.js:98:27)
    at CdpPage.waitForNavigation (/app/node_modules/puppeteer-core/lib/cjs/puppeteer/api/Page.js:619:43)
    at /app/scrapeAmazon.js:118:24
```

When I manually enter credentials into Chromium, everything is fine, so I don't think that's the issue.
```
AMZ_LOGIN=*******************
AMZ_PASS=*******************
AMZ_SECRET=****************************************************************************
HA_WEBHOOK_URL=https://*******************/api/webhook/*******************
Amazon_Sign_in_URL=https://www.amazon.de/ap/signin?openid.pape.max_auth_age=0&openid.return_to=https%3A%2F%2Fwww.amazon.de%2F%3Fref_%3Dnav_custrec_signin&openid.identity=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.assoc_handle=deflex&openid.mode=checkid_setup&openid.claimed_id=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.ns=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0
Amazon_Shopping_List_Page=https://www.amazon.de/alexaquantum/sp/alexaShoppingList?ref_=list_d_wl_ys_list_1
```

**Comment by thiagobruch** [comment]
Hi @stalsma ,

It looks like the sing-in page (at least testing from here) is the first option in the "IF" in the code.
```
	if (await page.$('#ap_password')) {
//          console.log('Element #ap_password found!');
            await page.type('#ap_email', amz_login);
            await page.type('#ap_password', amz_password);
            await page.click('#signInSubmit');
            await page.waitForNavigation();
	} else {
```

Can you try to add the console.log in this section?

**Comment by stalsma** [comment]
Prior to posting, I loaded up the script w a bunch of console.log() statements to isolate where the issue was.  I can confirm that I was definitely executing the else block.

Looking into this further, there seems to be a mismatch in the field IDs.  I configured the base version [amazon_shopping_list](https://github.com/jtbnz/amazon_shopping_list)) w/o issue.  After setting it up for my credentials and German URL, the login runs w/o an issue.  

That version of the script uses different IDs for the submit buttons.  But when I view the manually check the button ids in a browser, they match what is in your script.

Is there multiple correct urls for login, that use different button IDs?

**Comment by stalsma** [comment]
I recreated the container, and uncommented the list of ids that get found.
```
Element #ap_password not found. Retrying...
set login successfully
IDs found on the page: [
  'a-page',
  'authportal-center-section',
  'authportal-main-section',
  'auth-cookie-warning-message',
  'passkey-error-alert',
  'passkey-client-error-message',
  'passkey-generic-server-error-message',
  'passkey-server-error-message',
  'auth-email-claim',
  'ap_change_login_claim',
  'ap_email',
  'ap-credential-autofill-hint',
  'auth-fpp-link-bottom',
  'passwordErrorAnnounce',
  'ap_password',
  'auth-password-missing-alert',
  'auth-signin-button',
  'signInSubmit',
  'auth-signin-button-announce',
  'kmsi-weblab-treatment',
  'remember_me_learn_more_link',
  'ap-signin-alternative-option-divider',
  'auth-signin-via-passkey-section',
  'auth-signin-via-passkey-btn',
  'continue',

  'right-2',
  'auth-external-javascript',
  'be',
  'a-popover-root'
]
```

**auth-signin-button** (a span) is present as is ##signInSubmit## (the button).

I've added a few other mechanisms to submit the form, but no love.  I grabbed a screenshot of the form after adding the password.  The field is blank, but maybe that's how puppeteer behaves.

**Comment by thiagobruch** [comment]
Hi @stalsma ,

I just tried to access the sing-in URL from my end (Im in US and tried the german URL that you mentioned above) and I got the page with the captcha. Once I tried again, it went straight to the sign-in page.

In the code below that is part of the scraperAmazon.js, it is supposed to try to open the page, delay a bit and try again.
This used to make the captcha page go away.

Can you check if you have that in your code?

```
    // Navigate to Amazon login page
    await page.goto('https://www.amazon.com/ap/signin?openid.pape.max_auth_age=3600&openid.return_to=https%3A%2F%2Fwww.amazon.com%2Falexaquantum%2Fsp%2FalexaShoppingList%3Fref_%3Dlist_d_wl_ys_list_1&openid.identity=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.assoc_handle=amzn_alexa_quantum_us&openid.mode=checkid_setup&language=en_US&openid.claimed_id=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.ns=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0');
    sleep(1500, function() {
    // delay
    });
    await page.goto('https://www.amazon.com/ap/signin?openid.pape.max_auth_age=3600&openid.return_to=https%3A%2F%2Fwww.amazon.com%2Falexaquantum%2Fsp%2FalexaShoppingList%3Fref_%3Dlist_d_wl_ys_list_1&openid.identity=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.assoc_handle=amzn_alexa_quantum_us&openid.mode=checkid_setup&language=en_US&openid.claimed_id=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.ns=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0');
    //await page.goto(amz_signin_url);	
    // Enter username
```

**Comment by stalsma** [comment]
Hmmm.  I'm never seen a captcha.  Maybe it has something to do w/ IP location (IP's outside Germany get an extra challenge?)?

I vastly simplified the script, since it was working fine with the original project.  I essentially removed the initial loop, and a lot of the extra safety that was built-in.  And it still didn't work.  So I started looking at the config, and I think I found the actual issue.  By removing the stealth plugin (simply commented it out), I was able to get it working w/o any other modification:
![image](https://github.com/user-attachments/assets/2f0783a1-de75-4e9e-af57-bea836de3c28)

The Docker image I pulled (`thiagobruch/amazon-scrape-tb:latest`) uses the stealth code, but it isn't present in the github version of the file (`rootfs/app/scrapeAmazon.js`).
![image](https://github.com/user-attachments/assets/64616284-da4c-40e7-80fe-a65bb2f3eab0)

Is something out of sync?

**Comment by thiagobruch** [comment]
@stalsma you are correct.
I was working with Addon most of the time and when I updated the Docker version, I missed the stealth.
I will change that shortly to add the same code as the addon.

The stealth was one of the ways to get it to bypass the captcha.
The other was the workaround that I added :
1 - Open https://www.amazon.com
2 - Go to the sign-in URL: https://www.amazon.com/ap/signin?openid.pape.max_.....

However, when in a different country, that does not prevent the captcha sometimes, becuase you would need to go to https://www.amazon.de and then https://www.amazon.de/ap/signin?openid.pape.max_....

So I added the following to the code:
```
const url = amz_signin_url;
const parts = url.split('/');
const result = parts.slice(0, 3).join('/');
await page.goto(result, { waitUntil: 'load', timeout: 60000 });
```

**Comment by stalsma** [comment]
That explains it.  So I tried to deploy a new container based on the github master branch.    Simply checked it out and tried to build:
```
docker build -t amazon-scrape-tb .

[+] Building 0.8s (14/14) FINISHED                                                                                                                                                                    docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                                                  0.0s
 => => transferring dockerfile: 1.43kB                                                                                                                                                                                0.0s 
 => [internal] load metadata for docker.io/library/debian:stable-slim                                                                                                                                                 0.6s 
 => [internal] load .dockerignore                                                                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                                                                       0.0s 
 => [internal] load build context                                                                                                                                                                                     0.0s 
 => => transferring context: 229B                                                                                                                                                                                     0.0s 
 => [1/9] FROM docker.io/library/debian:stable-slim@sha256:382967fd7c35a0899ca3146b0b73d0791478fba2f71020c7aa8c27e3a4f26672                                                                                           0.0s 
 => CACHED [2/9] RUN apt-get update && apt-get install -y sudo curl &&     curl -fsSL https://deb.nodesource.com/setup_21.x | sudo -E bash - &&     apt-get install -y nodejs                                         0.0s 
 => CACHED [3/9] RUN sudo apt install libgl1-mesa-dri libgbm1 -y                                                                                                                                                      0.0s 
 => CACHED [4/9] RUN sudo sh -c 'echo "deb http://deb.debian.org/debian/ unstable main contrib non-free" >> /etc/apt/sources.list.d/debian.list'                                                                      0.0s 
 => CACHED [5/9] RUN sudo apt update && sudo apt install -y firefox --no-install-recommends                                                                                                                           0.0s 
 => CACHED [6/9] WORKDIR /app                                                                                                                                                                                         0.0s 
 => CACHED [7/9] COPY rootfs /                                                                                                                                                                                        0.0s 
 => CACHED [8/9] RUN npm install puppeteer otpauth dotenv                                                                                                                                                             0.0s 
 => CACHED [9/9] RUN chmod a+x /app/script.sh                                                                                                                                                                         0.0s 
 => exporting to image                                                                                                                                                                                                0.0s 
 => => exporting layers                                                                                                                                                                                               0.0s 
 => => writing image sha256:d3f1180c604be9af2520a08c1ef62a0fc189c2ef545572df540e380c83a53f02                                                                                                                          0.0s 
 => => naming to docker.io/library/amazon-scrape-tb      
```

But it errors out when I run it:
```
2024-08-23 12:08:48 exec /app/script.sh: no such file or directory
```
I tried building it in from CMD, PowerShell, and 

FWIW, I'm running this from Win 11.

**Comment by thiagobruch** [comment]
Hi @stalsma,

I've seen that error before, where the wrong interpreter (i.e. #!/bin/bash/) would cause the error "no such file or directory"  (it took me forever to figure that out because I could see that the file was there).

Let me try to rebuild that and see if I'm able to get it working on my environment.
You mentioned you are running on Win11, is that x64?

**Comment by stalsma** [comment]
Correct. But I didn't think that mattered. The error was in the shell of
the container.

**Comment by thiagobruch** [comment]
Hi @stalsma 

I included some screenshots for the Addon version today to show in which page the browser is connecting to.
I will add that to docker version here and let you know once it is ready.

**Comment by stalsma** [comment]
I'll keep an eye on this this ticket.  Right now, I have it working w/ the commented out Stealth Mode, but that is obviously not trivial to update.

I'll open some a few new tickets RE: documentation.

**Comment by thiagobruch** [comment]
Hi @stalsma,

Could you share your scrapeAmazon.js file?
There is somebody else with a similar issue and I would like to test your solution.

If that works, I will add (or in this case removed the stealth mode :) ) to the code.

Thank you,

**Comment by stalsma** [comment]
Here it is:
```
require('dotenv').config();

////////////// change to stealth
//const puppeteer = require('puppeteer');

// puppeteer-extra is a drop-in replacement for puppeteer,
// it augments the installed puppeteer with plugin functionality
const puppeteer = require('puppeteer-extra')

// add stealth plugin and use defaults (all evasion techniques)
// const StealthPlugin = require('puppeteer-extra-plugin-stealth')
// puppeteer.use(StealthPlugin())

//////////// end change to stealth

const OTPAuth = require('otpauth');  // For handling OTP
const fs = require('fs');

function getEnvVariable(key) {
    return process.env[key];
}

// Replace this with your actual secret key you get from the amazon add MFA page - and remove the spaces
const secret = getEnvVariable('AMZ_SECRET');
const amz_login = getEnvVariable('AMZ_LOGIN');
const amz_password = getEnvVariable('AMZ_PASS');
const delete_after_download = getEnvVariable('DELETE_AFTER_DOWNLOAD');
const log_level = getEnvVariable('log_level');
const amz_signin_url = getEnvVariable('Amazon_Sign_in_URL');
const amz_shoppinglist_url = getEnvVariable('Amazon_Shopping_List_Page');

// Create a new OTPAuth instance
const totp = new OTPAuth.TOTP({
  issuer: 'YourIssuer',
  label: amz_login,
  algorithm: 'SHA1',
  digits: 6,
  period: 30,
  secret: OTPAuth.Secret.fromBase32(secret)
});

// Generate OTP
const token = totp.generate();

//console.log(totp);
//console.log(token);

async function getOTP(secret) {
    const totp = new OTPAuth.TOTP({
        issuer: 'Amazon',
        label: 'Amazon OTP',
        algorithm: 'SHA1',
        digits: 6,
        period: 30,
        secret: secret
    });
    return totp.generate();
}

(async () => {
    const browser = await puppeteer.launch({
//            headless: true,
            defaultViewport: null,
            userDataDir: './tmp',
            args: [
        '--headless',
        '--no-sandbox',
        '--disable-setuid-sandbox',
//      '--single-process',
        '--disable-extensions',
        '--disable-gpu',
        '--disable-dev-shm-usage',
        '--disable-features=site-per-process'
                ],
//            args: ['--no-sandbox', '--disable-setuid-sandbox', '--single-process'],
//            product: 'firefox',
            executablePath: '/usr/bin/chromium',
//            executablePath: '/usr/bin/firefox',
//        dumpio: true,
          });

    const page = await browser.newPage();
        page.setDefaultTimeout(60000); // 60 seconds

// start loop code
let elementExists = false;
do {
//    Navigate to Amazon login page
//    await page.goto('https://www.amazon.com/ap/signin?openid.pape.max_auth_age=3600&openid.return_to=https%3A%2F%2Fwww.amazon.com%2Falex>
    await page.goto('https://www.amazon.com/', { waitUntil: 'load', timeout: 60000 });
    sleep(1500, function() {
    // delay
    });
    //await page.goto('https://www.amazon.com/ap/signin?openid.pape.max_auth_age=3600&openid.return_to=https%3A%2F%2Fwww.amazon.com%2Falex>
    await page.goto(amz_signin_url);
    elementExists = await page.$('#ap_email') !== null;
} while (!elementExists);
/// end loop code
// Enter username

	if (await page.$('#ap_password')) {
          console.log('Element #ap_password found!');
            await page.type('#ap_email', amz_login);
            await page.type('#ap_password', amz_password);
            await page.click('#signInSubmit');
            await page.waitForNavigation();
	} else {
          console.log('Element #ap_password not found. Retrying...');
            await new Promise(resolve => setTimeout(resolve, 1000)); // 30 second delay
            await page.type('#ap_email', amz_login);
            await page.click('#continue');
	console.log('sent un');
            
await page.waitForNavigation();
//          const ids = await page.$$eval('[id]', elements => elements.map(el => el.id));
//	    console.log('IDs found on the page:', ids);
            await page.type('#ap_password', amz_password);
            await page.click('#signInSubmit');
	console.log('sent pwd');
            
await page.waitForNavigation();
	}

	console.log('dealing w opt');
    // Handle OTP (if required)
    if (await page.$('#auth-mfa-otpcode')) {
        await page.type('#auth-mfa-otpcode', token);
        await page.click('#auth-signin-button');
        await page.waitForNavigation();
    }

	console.log('getting shopping list');

    // Navigate to Alexa Shopping List page
    //await page.goto('https://www.amazon.com/alexaquantum/sp/alexaShoppingList?ref_=list_d_wl_ys_list_1', { waitUntil: 'load', timeout: 60000 });
    await page.goto(amz_shoppinglist_url, { waitUntil: 'load', timeout: 60000 });

    const pageContent = await page.content();
    sleep(3000, function() {
    // delay
    });

  let itemTitles = await page.$$eval(".virtual-list .item-title", items =>
    items.map(item => item.textContent.trim())
  );

  // Format each item as <listItem>
  let formattedItems = itemTitles.map(item => `${item}`);

  // Convert the array to JSON format
  let jsonFormattedItems = JSON.stringify(formattedItems, null, 2);

  if(delete_after_download == "true") {
      let delete_buttons = await page.$$eval(".item-actions-2 button", buttons =>
          buttons.forEach(button => button.click())
      );
  }

  
  // Save the JSON formatted list to default.htm
  const outputDir = '.';
  if (!fs.existsSync(outputDir)){
    fs.mkdirSync(outputDir, { recursive: true });
  }
  fs.writeFileSync(`${outputDir}/list_of_items.json`, jsonFormattedItems);
	

  // Display the JSON formatted list
  console.log(jsonFormattedItems);

  // Close the browser when done
    await browser.close();
})();
function sleep(time, callback) {
  var stop = new Date().getTime();
  while (new Date().getTime() < stop + time) {
    ;
  }
  callback();
}
```

**Comment by thiagobruch** [comment]
Hi @stalsma 

Thank you for the info.
I added the lines to the BETA Version of the Addon and once it's been tested, I will apply to the final version.

thank you.
